scalding-core/src/main/scala/com/twitter/scalding/estimation/memory/MemoryEstimatorStepStrategy.scala

```diff
@@ -0,0 +1,109 @@
+
+  implicit val estimatorMonoid = new FallbackEstimatorMonoid[MemoryEstimate]
+
+  /**
+   * Make memory estimate, possibly overriding explicitly-set memory settings,
+   * and save useful info (such as the default & estimate) in JobConf for
+   * later consumption.
+   *
+   * Called by Cascading at the start of each job step.
+   */
+  final override def apply(
+    flow: Flow[JobConf],
+    preds: JList[FlowStep[JobConf]],
+    step: FlowStep[JobConf]): Unit = {
+
+    if (skipMemoryEstimation(step)) {
+      LOG.info(s"Skipping memory estimation as ${Config.MemoryEstimators} is not set ")
+    } else {
+      estimate(flow, preds.asScala, step)
+    }
+  }
+
+  private def skipMemoryEstimation(step: FlowStep[JobConf]): Boolean =
+    step.getConfig.get(Config.MemoryEstimators) == null
+
+  private def estimate(
+    flow: Flow[JobConf],
+    preds: Seq[FlowStep[JobConf]],
+    step: FlowStep[JobConf]): Unit = {
+    val conf = step.getConfig
+
+    Option(conf.get(Config.MemoryEstimators)).foreach { clsNames =>
+
+      val clsLoader = Thread.currentThread.getContextClassLoader
+
+      val estimators = StringUtility.fastSplit(clsNames, ",")
+        .map(clsLoader.loadClass(_).newInstance.asInstanceOf[Estimator[MemoryEstimate]])
+      val combinedEstimator = Monoid.sum(estimators)
+
+      val info = FlowStrategyInfo(flow, preds, step)
+
+      // get memory estimate
+      val memoryEstimate: Option[MemoryEstimate] = combinedEstimator.estimate(info)
+
+      memoryEstimate match {
+        case Some(MemoryEstimate(Some(mapMem), Some(reduceMem))) =>
+          LOG.info(s"Overriding map Xmx memory to: $mapMem and reduce Xmx memory to $reduceMem")
+          setMapMemory(mapMem, conf)
+          setReduceMemory(reduceMem, conf)
+        case Some(MemoryEstimate(Some(mapMem), _)) =>
+          LOG.info(s"Overriding only map Xmx memory to: $mapMem")
+          setMapMemory(mapMem, conf)
+        case Some(MemoryEstimate(_, Some(reduceMem))) =>
+          LOG.info(s"Overriding only reduce Xmx memory to $reduceMem")
+          setReduceMemory(reduceMem, conf)
+        case _ => LOG.info("Memory estimators didn't calculate any value. Skipping setting memory overrides")
+      }
+    }
+  }
+
+  private def setMapMemory(mapMem: Long, conf: Configuration): Unit = {
+    conf.setLong(MemoryEstimatorConfig.originalMapMemory, conf.getLong(Config.MapMemory, 0L))
+
+    val mapContainerMem = mapMem * MemoryEstimatorConfig.XmxToMemoryScaleFactor
+    val schedulerIncrement = conf.getInt(MemoryEstimatorConfig.yarnSchedulerIncrementAllocationMB, 512)
+    val roundedMapContainerMem = roundUp(mapContainerMem, schedulerIncrement).toLong
+
+    conf.setLong(Config.MapMemory, roundedMapContainerMem)
+
+    val mapOpts = conf.get(Config.MapJavaOpts, "")
+    //remove existing xmx / xms
+    val mapOptsWithoutXm = mapOpts.split(" ").filterNot(s => s.startsWith("-Xmx") || s.startsWith("-Xms")).mkString(" ")
```

Comment:
Can you put this in a separate function so we can add a unit test? This sounds like the thing that will break in the future on some edge case.